File: core/src/main/java/y2k/joyreactor/services/requests/MessageListRequest.java
package y2k.joyreactor.services.requests;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import y2k.joyreactor.Message;
import y2k.joyreactor.http.HttpClient;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * Created by y2k on 11/17/15.
 */
public class MessageListRequest {

    public List<Message> getMessages() {
        return messages;
    }

    public String getNextPage() {
        return nextPage;
    }

    private List<Message> messages = new ArrayList<>();
    private String nextPage;

    public void execute(String page) throws Exception {
        String url = page != null ? page : "http://joyreactor.cc/private/list";
        Document document = HttpClient.getInstance().getDocument(url);

        for (Element s : document.select("div.messages_wr > div.article")) {
            Message m = new Message();
            m.userName = s.select("div.mess_from > a").text();
            m.userImage = new UserImageRequest(m.userName).execute();
            m.text = s.select("div.mess_text").text();
            m.date = new Date(1000 * Long.parseLong(s.select("span[data-time]").attr("data-time")));
            m.isMine = s.select("div.mess_reply").isEmpty();
            messages.add(m);
        }

        Element nextNode = document.select("a.next").first();
        nextPage = nextNode == null ? null : nextNode.absUrl("href");

        loadUserImages();
    }

    private void loadUserImages() {
        for (Message m : messages)
            m.userImage = new UserImageRequest(m.userName).execute();
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/requests/MessageListRequest.kt
package y2k.joyreactor.services.requests

import y2k.joyreactor.http.HttpClient
import y2k.joyreactor.model.Message
import java.util.*

/**
 * Created by y2k on 11/17/15.
 */
class MessageListRequest(
    private val httpClient: HttpClient,
    private val imageRequest: UserImageRequest) {

    fun getMessages(page: String?): Pair<List<Message>, String?> {
        val messages = ArrayList<Message>()

        val url = page ?: "http://joyreactor.cc/private/list"
        val document = httpClient.getDocument(url)

        for (s in document.select("div.messages_wr > div.article")) {
            val username = s.select("div.mess_from > a").text()
            messages.add(
                Message(
                    s.select("div.mess_text").text(),
                    Date(1000 * java.lang.Long.parseLong(s.select("span[data-time]").attr("data-time"))),
                    s.select("div.mess_reply").isEmpty(),
                    username,
                    imageRequest.execute(username)))
        }

        val nextNode = document.select("a.next").first()
        val nextPage = nextNode?.absUrl("href")

        return messages to nextPage
    }
}